src/plugins/intel_gpu/src/plugin/ops/multinomial.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#include "intel_gpu/plugin/program_builder.hpp"
+#include "intel_gpu/plugin/common_utils.hpp"
+
+#include "openvino/op/multinomial.hpp"
+
+#include "intel_gpu/primitives/activation.hpp"
+#include "intel_gpu/primitives/broadcast.hpp"
+#include "intel_gpu/primitives/cum_sum.hpp"
+#include "intel_gpu/primitives/data.hpp"
+#include "intel_gpu/primitives/eltwise.hpp"
+#include "intel_gpu/primitives/multinomial.hpp"
+#include "intel_gpu/primitives/random_uniform.hpp"
+#include "intel_gpu/primitives/slice.hpp"
+
+#include <memory>
+#include <cstring>
+
+namespace ov {
+namespace intel_gpu {
+namespace {
+
+template<typename T>
+cldnn::data CreateScalarDataPrimitive(ProgramBuilder& p, const cldnn::primitive_id& name, T value) {
+    auto mem = p.get_engine().allocate_memory(
+        cldnn::layout{element::from<T>(), cldnn::format::bfyx, {1, 1, 1, 1}}, false);
+    cldnn::mem_lock<int8_t> host_mem{mem, p.get_engine().get_service_stream()};
+    std::memcpy(host_mem.data(), &value, sizeof value);
+    return {name, mem};
+}
+
+cldnn::data CreateShapeDataPrimitive(ProgramBuilder& p, const cldnn::primitive_id& name, Shape& value) {
+    auto mem = p.get_engine().allocate_memory(
+        cldnn::layout{element::Type_t::i64, cldnn::format::bfyx, {1, 1, 1, static_cast<int>(value.size())}}, false);
+    cldnn::mem_lock<int8_t> host_mem{mem, p.get_engine().get_service_stream()};
+    std::vector<std::int64_t> shape {};
+    std::copy(value.begin(), value.end(), std::back_inserter(shape));
+    std::memcpy(host_mem.data(), shape.data(), sizeof(std::int64_t) * shape.size());
+    return {name, mem};
+}
+
+static void CreateMultinomialOp(ProgramBuilder& p, const std::shared_ptr<ov::op::v13::Multinomial>& op) {
+    validate_inputs_count(op, { 2 });
+    auto inputs = p.GetInputInfo(op);
+    auto input0_shape = op->get_input_shape(0);
+    auto output_shape = op->get_output_shape(0);
+
+    const std::int64_t last_axis = input0_shape.size() - 1;
+    cldnn::primitive_id cumsum {layer_type_name_ID(op) + "_cumsum"};
+    if (op->get_log_probs()) {
+        cldnn::primitive_id exp {layer_type_name_ID(op) + "_exp"};
+        cldnn::activation exp_prim {exp, inputs[0], cldnn::activation_func::exp};
+        p.add_primitive(*op, exp_prim);
+        cldnn::cum_sum cumsum_primitive{cumsum, exp, last_axis};
+        p.add_primitive(*op, cumsum_primitive);
+    } else {
+        cldnn::cum_sum cumsum_primitive{cumsum, inputs[0], last_axis};
+        p.add_primitive(*op, cumsum_primitive);
+    }
+
+    std::vector<std::int32_t> slice_shape;
+    std::copy(input0_shape.begin(), input0_shape.end(), std::back_inserter(slice_shape));
```

Comment:
can't we use shape_of primitive?
Current solution works only for static shape.
Also if it is only intending for static shape, I am wondering why we should create scalar nodes as external inputs. 
(we can just use them as constant values as is).
If we intending it to be used for dynamic shape, we should 
1) use shape_of to get the shape 
2) use PartialShape instead of cldnn::tensor 

BTW what was the merit of decomposing this operation? 
Can't we just impl the entire behavior as a kernel?

Reply:
We cannot use shape_of primitive to calculate a shape for slice due to the following reasons:
1. There is additional modification of the shape value in the [L65](https://github.com/openvinotoolkit/openvino/pull/20616/files/284af4e7bf58ab749f686c58a24ba61592e209bd#diff-d37fd03576a88dc6378b4186794a2dea48d4038a10aadc4b54ecb92d75c2c112R65)
2. Interface of slice primitive requires shape as [attribute](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_gpu/include/intel_gpu/primitives/slice.hpp#L22) in its constructor.

Reply:
The operation is inherently compound. It's composed of the following operations:

- CumSum
- Slice
- RandomUniform
- Multinomial kernel which combines results of the previous operators.

In sake of efforts and further operation support I decided to reuse already existing primitives. Most of the primitives support static shapes only. In order to support dynamic shapes we have either add it in all used primitives or implement our own versions of them which support dynamism.
I validated target models described in the original ticket [CVS-111261](https://jira.devtools.intel.com/browse/CVS-111261) and found that the second input is always constant, therefore static shape implementation should be sufficient for them.

Reply:
> We cannot use shape_of primitive to calculate a shape for slice due to the following reasons:
> 
> 1. There is additional modification of the shape value in the [L65](https://github.com/openvinotoolkit/openvino/pull/20616/files/284af4e7bf58ab749f686c58a24ba61592e209bd#diff-d37fd03576a88dc6378b4186794a2dea48d4038a10aadc4b54ecb92d75c2c112R65)
> 2. Interface of slice primitive requires shape as [attribute](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_gpu/include/intel_gpu/primitives/slice.hpp#L22) in its constructor.

Actually "slice" primitive is not being used for now. We use crop primitive for that purpose. 
And modification of the last dim of the shape can be done like 
1. crop [:-2] 
2. concat + constant 1

Reply:
Slice potentially can be substituted with crop, but the next nested operation is random_uniform which also require cldnn::tensor with output shape.

Reply:
> > We cannot use shape_of primitive to calculate a shape for slice due to the following reasons:
> > 
> > 1. There is additional modification of the shape value in the [L65](https://github.com/openvinotoolkit/openvino/pull/20616/files/284af4e7bf58ab749f686c58a24ba61592e209bd#diff-d37fd03576a88dc6378b4186794a2dea48d4038a10aadc4b54ecb92d75c2c112R65)
> > 2. Interface of slice primitive requires shape as [attribute](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_gpu/include/intel_gpu/primitives/slice.hpp#L22) in its constructor.
> 
> Actually "slice" primitive is not being used for now. We use crop primitive for that purpose. And modification of the last dim of the shape can be done like
> 
>     1. crop [:-2]
> 
>     2. concat + constant 1

I've tried using crop with negative values to avoid passing shape information, but the crop primitive doesn't work with negative borders. There is explicit check for this case in [crop_inst](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_gpu/src/graph/crop.cpp#L218). Exception message this check generates looks as follows:
```
Error has occured for: multinomial:Multinomial_61_crop
Right/bottom/upper borders sizes: [b:0, f:-2, x:0, y:0, z:1, w:0, u:0, v:0, g:0]
0 value sizes: [b:0, f:0, x:0, y:0, z:0, w:0, u:0, v:0, g:0]
All Right/bottom/upper borders dimensions should not be less than 0 value dimensions.
Mismatching dimensions: Feature
Invalid border size: negative
```

Reply:
@yeonbok I figured out how to avoid passing shapes directly. Slice primitive also supports negative steps, so that it's not required to calculate them in the op factory directly. Though, to do this I had to add modifications in the slice primitive output layout calculation. It's sufficient to use shape_of primitives for other cases.